### PR TITLE
Add asana integration test and intrument bugsnag on api call error

### PR DESCRIPTION
**Pull request by asasmoyo**
For https://github.com/toggl/backend/issues/9130

Bugsnag is needed since lots of unknown errors happening with asana

### CODE REVIEW

**Comment by asasmoyo** [comment]
rolled to production

File: asana_test.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+	"code.google.com/p/goauth2/oauth"
+)
+
+func createAsanaService() Service {
+	s := &AsanaService{}
+	token := oauth.Token{
+		AccessToken: os.Getenv("ASANA_PERSONAL_TOKEN"),
+	}
+	s.token = token
```
Comment (attilio-work) [inline asana_test.go]: Minor: you could just to `s.token = oauth.Token{...`

File: asana_test.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+		AccountID: numberStrToInt64(os.Getenv("ASANA_ACCOUNT_ID")),
+	}
+	return s
+}
+
+func TestAsanaAccounts(t *testing.T) {
+	s := createAsanaService()
+
+	accounts, err := s.Accounts()
+	if err != nil {
```
Comment (attilio-work) [inline asana_test.go]: It would be slightly better to use the assertions.

File: asana_test.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+		t.Error("error calling accounts(), err:", err)
+	}
+
+	if len(accounts) != 1 {
```
Comment (attilio-work) [inline asana_test.go]: Here I would check the account ID too.

File: asana_test.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+func TestAsanaProjects(t *testing.T) {
+	s := createAsanaService()
+
+	projects, err := s.Projects()
+	if err != nil {
+		t.Error("error calling projects(), err:", err)
+	}
+
+	if len(projects) <= 20 {
+		t.Error("should get more than 20 project, please create at least 20 project to test pagination")
```
Comment (attilio-work) [inline asana_test.go]: typo: project**s**

File: db/schema.sql [hunk 5]
```diff
@@ -53,19 +53,23 @@ CREATE UNIQUE INDEX pipes_queue_unique ON queued_pipes (workspace_id, key, coale
 CREATE OR REPLACE FUNCTION get_queued_pipes() RETURNS TABLE(workspace_id INTEGER, key VARCHAR(50)) AS $$
 BEGIN
   RETURN QUERY
-  UPDATE
-    queued_pipes
-  SET
-    locked_at = NOW()
-  FROM (
+  WITH available_queue AS (
```
Comment (attilio-work) [inline db/schema.sql]: How is this change related to the pagination?

File: handlers.go [hunk 6]
```diff
@@ -347,7 +360,11 @@ func postPipeRun(req Request) Response {
 		return badRequest(msg)
 	}
 	if pipe.ID == "users" {
-		go pipe.run()
+		go func() {
+			workspaceLock.Lock()
+			pipe.run()
+			workspaceLock.Unlock()
+		}()
 		time.Sleep(500 * time.Millisecond)
```
Comment (attilio-work) [inline handlers.go]: I know this is not part of your change, but...
Do you know why they added this sleep here? Do they expect the above go routine to always run in 500 ms?

File: handlers.go [hunk 7]
```diff
@@ -347,7 +360,11 @@ func postPipeRun(req Request) Response {
 		return badRequest(msg)
 	}
 	if pipe.ID == "users" {
-		go pipe.run()
+		go func() {
```
Comment (attilio-work) [inline handlers.go]: Also, this looks like a separate issue, not pagination.
What was exactly the problem here? What were the symptoms?

File: asana_test.go (hunk 1 above)
Comment (asasmoyo) [reply asana_test.go]: i have something even simpler :D

`s.token.AccessToken = os.Getenv("ASANA_PERSONAL_TOKEN")`

File: handlers.go (hunk 7 above)
Comment (asasmoyo) [reply handlers.go]: yes it is changes from separate issue. it's https://github.com/toggl/pipes-api/pull/40

current deployed pipes is this branch, so in order to not mess current code i did the pr on top of this branch

File: handlers.go (hunk 6 above)
Comment (asasmoyo) [reply handlers.go]: i don't have answer for this 😅 

but in reality, there always be the case when it takes more than 500ms. can be network problems or too many data to load so it will need to run multiple api calls to fetch them all

File: db/schema.sql (hunk 5 above)
Comment (asasmoyo) [reply db/schema.sql]: answered in https://github.com/toggl/pipes-api/pull/39#discussion_r355828062

File: asana_test.go (hunk 2 above)
Comment (asasmoyo) [reply asana_test.go]: could you please elaborate what 'assertions' you are referring to?

**Comment by asasmoyo** [comment]
yes, well it was combination of lots of complaints coming from users, no pipes in staging and review is taking time 😅

File: asana_test.go (hunk 2 above)
Comment (attilio-work) [reply asana_test.go]: I mean `check.v1` package and `c.Assert`, like in [`toggl_api`](https://github.com/toggl/toggl_api/blob/master/src/toggl_api/push_notification_test.go#L90)

**Comment by attilio-work** [comment]
looks ok to me now -- if you can introduce `check.v1` then great, if it is too much work, then issue can be closed

**Comment by asasmoyo** [comment]
yap would be nice but i personally don't like to do it in current code, would definitely do it on reworked pipes code :)

ok i'm closing this
